src/swap.rs

```diff
@@ -281,20 +404,54 @@ async fn hbit_herc20<A, B, SC>(
     secp: &bitcoin::secp256k1::Secp256k1<SC>,
 ) -> anyhow::Result<()>
 where
-    A: hbit::Fund + herc20::RedeemAsAlice,
-    B: herc20::Deploy + herc20::Fund + hbit::RedeemAsBob,
+    A: hbit::Fund + herc20::RedeemAsAlice + hbit::Refund + SafeToFund + SafeToRedeem,
+    B: herc20::Deploy + herc20::Fund + hbit::RedeemAsBob + herc20::Refund + SafeToFund,
     SC: bitcoin::secp256k1::Signing,
 {
+    if !alice.is_safe_to_fund(herc20_params.expiry).await? {
+        return Ok(());
+    }
+
     let hbit_funded = alice.fund(&hbit_params).await?;
 
+    if !bob.is_safe_to_fund(herc20_params.expiry).await? {
+        alice.refund(&hbit_params, hbit_funded, secp).await?;
```

Comment:
I am not completely sure why but it is a bit weird that we are asking Bob here before Alice does her refund.

I guess it is fine because it is the same process so the other actor is not gonna be malicious :)

Reply:
> Alice should refund as soon as she can refund, when it's past Alpha expiry.

This refund method waits until refund is possible and then refunds. Maybe the name could reflect that somehow.

>  Bob should fund before beta expiry, if beta expiry is passed Bob should refund or not fund.

I think that's what we're doing here. A source of confusion may be that we're using `is_safe_to_fund` for both deploy and fund, for simplicity. I can split them up.

> As we work with absolutely expiries there is no safe to fund notion for Bob.
safe to fund for bob sounds like Bob would try to fund after beta expiry but not too close to alpha expiry.

You are right, the name of this function does not work here. 

Here the logic is: if Beta has expired, end the protocol. For Alice because she cannot safely reveal the secret if Bob could refund just before and learn the secret for free; for Bob because the other party should not continue, so we may as well abort and not waste money on fees for a swap that should never succeed.

Maybe we could just use `{alpha, beta}_has_expired` instead of the not always correct `is_safe_to_fund`.

Reply:
Yes I'd be keen for a rename and just use expiries as time markers to decide.